src/Swift.Runtime/src/Swift/SwiftArray.cs

```diff
@@ -0,0 +1,285 @@
+    public static nuint ElementSize => _elementSize;
+    static TypeMetadata ISwiftObject.GetTypeMetadata()
+    {
+        return TypeMetadata.Cache.GetOrAdd(typeof(SwiftArray<Element>), _ => SwiftArrayPInvokes.PInvoke_getMetadata(TypeMetadataRequest.Complete, ElementTypeMetadata));
+    }
+
+    static TypeMetadata ElementTypeMetadata
+    {
+        get => TypeMetadata.GetTypeMetadataOrThrow<Element>();
+    }
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return new SwiftArray<Element>(handle);
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<SwiftArray<Element>>.GetTypeMetadata();
+        unsafe
+        {
+            fixed (void* _payloadPtr = &buffer)
+            {
+                metadata.ValueWitnessTable->InitializeWithCopy((void*)swiftDest, (void*)_payloadPtr, metadata);
+            }
+        }
+        return swiftDest;
+    }
+
+    /// <summary>
+    /// Gets the protocol conformance descriptor for the given type.
+    /// </summary>
+    /// <typeparam name="TProtocol"></typeparam>
+    /// <returns></returns>
+    static ProtocolConformanceDescriptor ISwiftObject.GetProtocolConformanceDescriptor<TProtocol>()
+        where TProtocol : class
+    {
+        return ProtocolConformanceDescriptor.Zero;
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftArray from the given handle.
+    /// </summary>
+    unsafe SwiftArray(SwiftHandle handle)
+    {
+        this.buffer = *(ArrayBuffer*)(handle);
+    }
+
+    /// <summary>
+    /// Constructs a new empty SwiftArray.
+    /// </summary>
+    public SwiftArray()
+    {
+        buffer = SwiftArrayPInvokes.Init(ElementTypeMetadata);
+    }
+
+    /// <summary>
+    /// Gets the number of elements in the array.
+    /// </summary>
+    public unsafe int Count
+    {
+        get
+        {
+            return (int)SwiftArrayPInvokes.Count(buffer, ElementTypeMetadata);
+        }
+    }
+
+    /// <summary>
+    /// Appends the given element to the array.
+    /// </summary>
+    public unsafe void Append(Element item)
+    {
+        IntPtr T0Payload = IntPtr.Zero;
+        var metadata = SwiftObjectHelper<SwiftArray<Element>>.GetTypeMetadata();
+        try
+        {
+            T0Payload = (IntPtr)NativeMemory.Alloc(ElementSize);
+            SwiftMarshal.MarshalToSwift(item, (IntPtr)T0Payload);
+
+            fixed (void* _payloadPtr = &buffer)
+            {
+                SwiftArrayPInvokes.Append(T0Payload, metadata, new SwiftSelf(_payloadPtr));
+            }
+        }
+        finally
+        {
+            NativeMemory.Free((void*)T0Payload);
+        }
+    }
+
+    /// <summary>
+    /// Inserts the given element at the given index.
+    /// </summary>
+    public unsafe void Insert(int index, Element item)
+    {
+        if (index < 0 || index >= Count)
+            throw new ArgumentOutOfRangeException(nameof(index));
```

Comment:
Just curious about our general approach: Technically there's nothing in this projection which requires the index to be in the bounds of the array. And for other projections we won't be able to generate argument validation this way. Should we leave this up to the Swift's implementation fully? (For this case, I could imagine that Insert after end means - expand the array and set the element for example).

Reply:
Removed, not necessary.